src/Nethermind/Nethermind.Synchronization/SyncServer.cs

```diff
@@ -181,34 +181,41 @@ public void AddNewBlock(Block block, ISyncPeer nodeWhoSentTheBlock)
 
             bool isBlockBeforeTheSyncPivot = block.Number < _pivotNumber;
             bool isBlockOlderThanMaxReorgAllows = block.Number < (_blockTree.Head?.Number ?? 0) - Sync.MaxReorgLength;
-            bool isBlockTotalDifficultyLow = block.TotalDifficulty < _blockTree.BestSuggestedHeader.TotalDifficulty
-                                             && (_specProvider.TerminalTotalDifficulty == null || block.TotalDifficulty < _specProvider.TerminalTotalDifficulty); // terminal blocks with lower TTD might be useful for smooth merge transition
-            if (isBlockBeforeTheSyncPivot || isBlockTotalDifficultyLow || isBlockOlderThanMaxReorgAllows) return;
 
-            lock (_recentlySuggested)
+            if (isBlockBeforeTheSyncPivot || isBlockOlderThanMaxReorgAllows)
             {
-                if (_recentlySuggested.Get(block.Hash) != null) return;
-                _recentlySuggested.Set(block.Hash, _dummyValue);
+                return;
             }
 
-            if (_specProvider.TerminalTotalDifficulty != null && block.TotalDifficulty >= _specProvider.TerminalTotalDifficulty)
+            if (_recentlySuggested.Set(block.Hash))
             {
-                if (_logger.IsInfo) _logger.Info($"Peer {nodeWhoSentTheBlock} sent block {block} with total difficulty {block.TotalDifficulty} higher than TTD {_specProvider.TerminalTotalDifficulty}");
-            }
+                if (_specProvider.TerminalTotalDifficulty != null && block.TotalDifficulty >= _specProvider.TerminalTotalDifficulty)
+                {
+                    if (_logger.IsInfo) _logger.Info($"Peer {nodeWhoSentTheBlock} sent block {block} with total difficulty {block.TotalDifficulty} higher than TTD {_specProvider.TerminalTotalDifficulty}");
+                }
 
-            ValidateSeal(block, nodeWhoSentTheBlock);
-            if ((_syncModeSelector.Current & (SyncMode.FastSync | SyncMode.StateNodes)) == SyncMode.None
-                || (_syncModeSelector.Current & SyncMode.Full) != SyncMode.None)
-            {
-                LogBlockAuthorNicely(block, nodeWhoSentTheBlock);
-                SyncBlock(block, nodeWhoSentTheBlock);
+                ValidateSeal(block, nodeWhoSentTheBlock);
+                SyncMode syncMode = _syncModeSelector.Current;
+                bool notInFastSyncNorStateSync = (syncMode & (SyncMode.FastSync | SyncMode.StateNodes)) == SyncMode.None;
+                bool inFullSyncOrWaitingForBlocks = (syncMode & (SyncMode.Full | SyncMode.WaitingForBlock)) != SyncMode.None;
+                if (notInFastSyncNorStateSync || inFullSyncOrWaitingForBlocks)
+                {
+                    LogBlockAuthorNicely(block, nodeWhoSentTheBlock);
+                    SyncBlock(block, nodeWhoSentTheBlock);
+                }
+
+                // we might null total difficulty for a block in a block tree as we don't trust the message
+                Block blockToBroadCast = block.TotalDifficulty is null
```

Comment:
so we're rebroadcasting block with potentially wrong difficulty. What if in SyncBlock headerValidator rejected block? Will we broadcast invalid block?

Reply:
If `ValidateSeal` or `SyncBlock` reject the block then they throw an exception, so we don't rebroadcast the block. Let me think if I can refactor it to make it clearer.

Reply:
but wait, imagine that we don't have the parent of this block. In this case we will rebrodcast it. Shouldn't we call BroadcastBlock in SyncBlock method?

Reply:
Refactored it